# 🔢 Datové typy a základní operátory

---

## Základní datové typy v Pythonu

Každá hodnota (a tím i každá proměnná) má v Pythonu svůj **datový typ**. Datový typ určuje, jaký druh dat proměnná uchovává a jaké operace s ní můžeme provádět.

### Dynamické typování

Python používá **dynamické typování** – to znamená, že:
- **Nemusíme** datový typ proměnné předem deklarovat (na rozdíl od jazyků jako C nebo Java)
- Python **automaticky rozpozná** typ podle přiřazené hodnoty
- Typ proměnné se **může v průběhu programu měnit** – stačí do ní uložit hodnotu jiného typu

```python
x = 42          # x je int (celé číslo)
x = "Ahoj"     # teď je x str (text) – typ se změnil!
```

### Přehled základních datových typů

| Datový typ | Python název | Příklady hodnot | Popis |
|---|---|---|---|
| **Celé číslo** | `int` (integer) | `0`, `42`, `-7`, `1000000` | Kladná i záporná celá čísla bez desetinné části |
| **Desetinné číslo** | `float` | `3.14`, `-0.5`, `2.0`, `1e10` | Čísla s desetinnou **tečkou** (ne čárkou!) |
| **Text (řetězec)** | `str` (string) | `"Ahoj"`, `'Python'`, `""` | Libovolný text v uvozovkách nebo apostrofech |
| **Logická hodnota** | `bool` (boolean) | `True`, `False` | Pouze dvě možné hodnoty – pravda/nepravda |

> ⚠️ **Pozor na desetinnou tečku!** V Pythonu (a programování obecně) se **nepoužívá čárka** jako v češtině, ale **tečka**: `3.14` (ne `3,14`).

> 💡 **Zajímavost:** Python podporuje i další datové typy jako `list` (seznam), `dict` (slovník), `tuple` (n-tice) aj. – o nich se dozvíte v pozdějších lekcích.

---

## 🔍 Funkce `type()` – zjištění datového typu

Pokud si nejste jisti, jaký datový typ má proměnná, můžete použít vestavěnou funkci `type()`. Vrací informaci o datovém typu hodnoty.

```python
type(hodnota)        # vrátí datový typ dané hodnoty
type(promenna)       # vrátí datový typ hodnoty uložené v proměnné
```

Výstup bude ve formátu `<class 'typ'>` – slovo `class` (třída) zatím neřešte, důležitý je název typu za ním.

In [ ]:
# Vytvoříme proměnné různých datových typů
cele_cislo = 34                # int
jmeno = "Honza"                # str
logicka_hodnota = False        # bool
desetinne_cislo = 56.981       # float

# Vypíšeme hodnotu i její datový typ
print(cele_cislo, "→", type(cele_cislo))
print(jmeno, "→", type(jmeno))
print(logicka_hodnota, "→", type(logicka_hodnota))
print(desetinne_cislo, "→", type(desetinne_cislo))

# Dynamické typování – proměnná může změnit typ
print()
x = 10
print("x =", x, "→", type(x))   # int
x = "teď jsem text"
print("x =", x, "→", type(x))   # str

34 <class 'int'>
Honza <class 'str'>
False <class 'bool'>
56.981 <class 'float'>


> 💡 Výstup `<class 'int'>` znamená, že hodnota patří do třídy (class) `int`. Pojem **třída** podrobně probereme v lekci o objektově orientovaném programování.

---

## 🔄 Přetypování (konverze datových typů)

Někdy potřebujeme **převést hodnotu z jednoho datového typu na jiný**. Tomuto procesu se říká **přetypování** (type casting / type conversion).

### Funkce pro přetypování

| Funkce | Co dělá | Příklad |
|---|---|---|
| `int()` | Převede na celé číslo | `int("42")` → `42`, `int(3.9)` → `3` (ořízne desetinnou část!) |
| `float()` | Převede na desetinné číslo | `float("3.14")` → `3.14`, `float(5)` → `5.0` |
| `str()` | Převede na text | `str(42)` → `"42"`, `str(True)` → `"True"` |
| `bool()` | Převede na logickou hodnotu | `bool(0)` → `False`, `bool(1)` → `True`, `bool("")` → `False` |

### Kdy přetypování potřebujeme?

Nejčastěji při **vstupu od uživatele** – funkce `input()` **vždy vrací `str`** (text), i když uživatel zadá číslo. Pokud chceme s číslem počítat, musíme ho nejprve převést.

> ⚠️ **Pozor:** Převod `int("3.14")` vyhodí chybu! Nelze přímo převést desetinný text na celé číslo. Musíte nejprve převést na `float` a pak na `int`: `int(float("3.14"))`.

In [ ]:
# === Ukázky přetypování ===

# int → float
cislo_int = 64
cislo_float = float(cislo_int)
print(cislo_int, type(cislo_int), "→", cislo_float, type(cislo_float))

# int → str → int (tam a zpět)
cislo = 12
retezec = str(cislo)
zpet_cislo = int(retezec)
print(cislo, type(cislo), "→", retezec, type(retezec), "→", zpet_cislo, type(zpet_cislo))

# float → int (pozor: ořízne desetinnou část, nezaokrouhluje!)
print()
print("float → int:")
print(int(3.9))    # 3 (ne 4!)
print(int(3.1))    # 3
print(int(-2.7))   # -2

# Převod na bool
print()
print("Převod na bool:")
print(bool(0))      # False – nula je "nepravda"
print(bool(42))     # True – jakékoli nenulové číslo je "pravda"
print(bool(""))     # False – prázdný řetězec je "nepravda"
print(bool("text")) # True – neprázdný řetězec je "pravda"

64 <class 'int'> 64.0 <class 'float'>
12 <class 'int'> 12 <class 'str'> 12 <class 'int'>


### Praktické použití: `input()` + přetypování

Jak víme, `input()` vždy vrací `str`. Pro práci s čísly je nutné použít přetypování:

In [ ]:
# Přetypování vstupu na celé číslo
moje_cislo = int(input("Zadej celé číslo: "))
print("Zadal jsi:", moje_cislo, "| typ:", type(moje_cislo))

# Přetypování vstupu na desetinné číslo
desetinne = float(input("Zadej desetinné číslo: "))
print("Zadal jsi:", desetinne, "| typ:", type(desetinne))

---

# ➕ Operátory

## Co je to operátor?

**Operátor** je speciální symbol, který provádí určitou operaci nad jednou nebo více hodnotami (operandy).

```python
vysledek = 5 + 3
#          ↑ ↑ ↑
#     operand | operand
#          operátor
```

### Důležité pravidlo
Chování operátoru **závisí na datovém typu** operandů! Stejný operátor může s různými typy dělat různé věci (nebo vyhodit chybu).

---

## Přehled aritmetických operátorů

| Operátor | Operace | Příklad | Výsledek |
|:---:|---|---|---|
| `+` | Sčítání / spojení | `5 + 3` / `"A" + "B"` | `8` / `"AB"` |
| `-` | Odčítání | `5 - 3` | `2` |
| `*` | Násobení / opakování | `5 * 3` / `"A" * 3` | `15` / `"AAA"` |
| `/` | Dělení (vždy `float`) | `7 / 2` | `3.5` |
| `//` | Celočíselné dělení | `7 // 2` | `3` |
| `%` | Modulo (zbytek po dělení) | `7 % 2` | `1` |
| `**` | Mocnění | `2 ** 3` | `8` |

---

## Operátor přiřazení `=`

Operátor `=` **není rovnost** jako v matematice! V programování znamená **„ulož hodnotu"**.

**Čteme zprava doleva:** *Vezmi hodnotu vpravo a ulož ji do proměnné vlevo.*

```python
x = 5       # ulož číslo 5 do proměnné x
y = x       # ulož HODNOTU proměnné x (tedy 5) do proměnné y
x = x + 1   # vezmi hodnotu x (5), přičti 1, výsledek (6) ulož zpět do x
```

> 🤔 **Zamyslete se:** V matematice zápis `x = x + 1` nedává smysl (žádné číslo se nerovná sobě + 1). V programování je to ale naprosto běžný a důležitý zápis!

---

## Sčítání `+` a odčítání `-`

### S čísly (`int`, `float`)
Operátory `+` a `-` provádějí klasické **sčítání a odčítání**.

### S textem (`str`) – operátor `+`
Operátor `+` u řetězců provádí **konkatenaci** (spojení dvou textů dohromady). Odčítání textů v Pythonu **není možné** a vyvolá chybu `TypeError`.

```python
"Ahoj" + " " + "světe"   # → "Ahoj světe"
"Ahoj" - "A"              # ❌ TypeError!
```

> 💡 Při spojování textů se **nepřidává mezera automaticky** – musíte ji přidat sami: `"Ahoj" + " " + "světe"`

In [ ]:
# === Sčítání a odčítání čísel ===
a = 15
b = 37
print("Sčítání:", a, "+", b, "=", a + b)
print("Odčítání:", a, "-", b, "=", a - b)

# === Konkatenace (spojení) řetězců ===
print()
jmeno = "Ahoj"
prijmeni = "Jarda"
cele = jmeno + " " + prijmeni   # spojíme s mezerou uprostřed
print("Spojení:", '"' + jmeno + '"', "+", '"' + prijmeni + '"', "=", '"' + cele + '"')

# === Odčítání řetězců NELZE – odkomentujte pro chybu ===
# vysledek = jmeno - prijmeni   # TypeError: unsupported operand type(s)

---

## Násobení `*`

### S čísly
Klasické **násobení** dvou čísel.

### S textem a číslem
Operátor `*` umožňuje **opakování řetězce** – vynásobí text celým číslem a vytvoří opakovanou kopii.

```python
"Ha" * 3      # → "HaHaHa"
3 * "Bum "    # → "Bum Bum Bum "
```

> ⚠️ Násobení dvou řetězců (`"Ahoj" * "svět"`) **není možné** – pouze text × celé číslo.

In [ ]:
# === Násobení čísel ===
a = 15
b = 37
print("Násobení:", a, "*", b, "=", a * b)

# === Opakování řetězce ===
print()
napis = "Ha"
print("Opakování:", '"' + napis + '"', "* 3 =", '"' + napis * 3 + '"')
print("Oddělovač:", "-" * 30)

# Praktické využití – formátování výstupu
print()
print("=" * 20)
print("  VÝSLEDKY TESTU  ")
print("=" * 20)

---

## Dělení `/`

Operátor `/` provádí klasické **dělení**. 

### Důležitá vlastnost
Výsledek dělení je **vždy `float`** (desetinné číslo), i když dělíte dvě celá čísla a výsledek je „celý":

```python
6 / 3     # → 2.0 (ne 2!)
7 / 2     # → 3.5
10 / 10   # → 1.0
```

> ⚠️ **Dělení nulou** (`x / 0`) vyvolá chybu `ZeroDivisionError`!

In [ ]:
# === Dělení celých čísel – výsledek je vždy float! ===
a = 15
b = 3
print(a, "/", b, "=", a / b, "→ typ:", type(a / b))

a = 7
b = 2
print(a, "/", b, "=", a / b, "→ typ:", type(a / b))

# === Dělení desetinných čísel ===
a = 15.5
b = 3.1
print(a, "/", b, "=", a / b, "→ typ:", type(a / b))

# Dělení nulou – odkomentujte pro chybu:
# print(10 / 0)   # ZeroDivisionError!

---

## Celočíselné dělení `//` a modulo `%`

Tyto dva operátory spolu úzce souvisí – vracejí dvě části výsledku dělení:

### Celočíselné dělení `//`
Vrátí **celou část** výsledku dělení (zaokrouhlí dolů).
- Lze si představit jako: *„Kolikrát se dělitel vejde do dělence?"*

### Modulo `%`
Vrátí **zbytek** po celočíselném dělení.
- Lze si představit jako: *„Co zbyde po celočíselném dělení?"*

### Vizuální příklad

```
15 ÷ 4 = 3 (zbytek 3)

15 // 4 = 3    ← celá část
15  % 4 = 3    ← zbytek
```

### Ověření

Platí vždy: **`(a // b) * b + (a % b) == a`**

```python
a = 15, b = 4:   (15 // 4) * 4 + (15 % 4)  =  3 * 4 + 3  =  15 ✅
```

### Datový typ výsledku
- Pokud oba operandy jsou `int` → výsledek je `int`
- Pokud je alespoň jeden `float` → výsledek je `float`

> 💡 **Kde se modulo hodí?** Například pro zjištění, zda je číslo sudé: `cislo % 2 == 0`

In [ ]:
# === Celočíselné dělení a modulo s int ===
a = 15
b = 4
print(f"{a} // {b} = {a // b}")     # celá část: 3
print(f"{a}  % {b} = {a % b}")      # zbytek: 3
print(f"Ověření: {a // b} * {b} + {a % b} = {(a // b) * b + (a % b)}")  # = 15

# === S float – pozor na nepřesnost! ===
print()
a = 15.5
b = 3.1
print(f"{a} // {b} = {a // b}")     # celá část (float) → 4.0
print(f"{a}  % {b} = {a % b}")      # zbytek → NE přesně 3.1, ale 3.0999999999999996 (!)
# ↑ Proč? Vysvětlení v buňce níže.

# === Praktické použití: sudé/liché číslo ===
print()
cislo = 7
print(f"Je {cislo} sudé?", cislo % 2 == 0)   # False (7 je liché)
cislo = 12
print(f"Je {cislo} sudé?", cislo % 2 == 0)   # True (12 je sudé)

Integer: 15 // 3 = 5 zbytek 0 <class 'int'> <class 'int'>
Float: 15.5 // 3.1 = 4.0 zbytek 3.0999999999999996 <class 'float'> <class 'float'>


---

### ⚠️ Proč `15.5 % 3.1` není přesně `3.1`? — Nepřesnost desetinných čísel

Výsledek `3.0999999999999996` místo očekávaného `3.1` **není chyba Pythonu** – je to důsledek toho, jak počítače interně ukládají desetinná čísla.

#### Jak počítač ukládá `float`?

Počítač pracuje ve **dvojkové soustavě** (0 a 1). Některá desetinná čísla **nelze ve dvojkové soustavě vyjádřit přesně** – podobně jako v desítkové soustavě nelze přesně zapsat $\frac{1}{3} = 0.33333...$

Číslo `3.1` je ve dvojkové soustavě nekonečný periodický zlomek. Počítač ho uloží s **malou zaokrouhlovací chybou** (řádově $10^{-16}$). Tato chyba se pak projeví ve výpočtech.

```python
print(0.1 + 0.2)          # → 0.30000000000000004 (ne 0.3!)
print(0.1 + 0.2 == 0.3)   # → False (!!)
```

#### Jak se s tím vypořádat?

| Řešení | Příklad |
|---|---|
| **Zaokrouhlení** pomocí `round()` | `round(0.1 + 0.2, 1)` → `0.3` |
| **Porovnání s tolerancí** | `abs(vysledek - ocekavany) < 0.0001` |
| **Pro přesné výpočty** (finance apod.) | modul `decimal` |

> 💡 Toto není problém jen Pythonu – stejná nepřesnost existuje ve **všech** programovacích jazycích, které používají IEEE 754 standard (C, Java, JavaScript, …).

In [ ]:
# === Ukázka nepřesnosti float ===
print("0.1 + 0.2 =", 0.1 + 0.2)              # 0.30000000000000004
print("0.1 + 0.2 == 0.3?", 0.1 + 0.2 == 0.3)  # False!

# === Řešení: zaokrouhlení ===
print()
print("round(0.1 + 0.2, 1) =", round(0.1 + 0.2, 1))   # 0.3
print("round(15.5 % 3.1, 1) =", round(15.5 % 3.1, 1))  # 3.1

---

## Mocnění `**`

Operátor `**` provádí **umocnění** (exponentiation): základ umocněný na exponent.

```python
zaklad ** exponent
2 ** 3    # → 8  (= 2³ = 2 × 2 × 2)
```

### Datový typ výsledku
- `int ** int` → `int` (pokud je exponent nezáporný)
- Pokud je některý operand `float` → výsledek je `float`

### Speciální případy
| Výraz | Výsledek | Poznámka |
|---|---|---|
| `x ** 0` | `1` | Cokoliv na nultou je 1 |
| `x ** 1` | `x` | Cokoliv na první je samo sebe |
| `x ** 0.5` | $\sqrt{x}$ | Odmocnina! $x^{0.5} = \sqrt{x}$ |
| `x ** -1` | $\frac{1}{x}$ | Převrácená hodnota |

> 🤔 **Kontrolní otázka:** Co znamená `9 ** 0.5`? *(Odpověď: $ 9^{0.5} = \sqrt{9} = 3.0$)*

In [ ]:
# === Mocnění celých čísel ===
a = 2
b = 10
print(f"{a} ** {b} = {a ** b}")           # 1024
print(f"Typ: {type(a ** b)}")             # int

# === Mocnění s float ===
print()
a = 15.5
b = 3.1
print(f"{a} ** {b} = {a ** b}")           # float
print(f"Typ: {type(a ** b)}")

# === Speciální případy ===
print()
print(f"5 ** 0   = {5 ** 0}")             # 1 (cokoliv na nultou)
print(f"9 ** 0.5 = {9 ** 0.5}")           # 3.0 (odmocnina z 9)
print(f"2 ** -1  = {2 ** -1}")            # 0.5 (převrácená hodnota)
print(f"4 ** -0.5 = {4 ** -0.5}")         # 0.5 (1/√4)

### 💡 Operátory pracují i přímo s hodnotami

Hodnoty nemusí být uložené v proměnných – operátor lze použít přímo s **literály** (konkrétními hodnotami) nebo **kombinovat proměnné s hodnotami**:

In [ ]:
# Přímo s hodnotami (literály)
print(2 + 3)           # 5
print(10 ** 2)         # 100

# Kombinace proměnné a hodnoty
a = 5
print("Desetinásobek", a, "je", a * 10)

---

## ⚖️ Porovnávací (relační) operátory

Porovnávací operátory **porovnávají dvě hodnoty** a vrací vždy **`bool`** (`True` nebo `False`).

| Operátor | Význam | Příklad | Výsledek |
|:---:|---|---|---|
| `==` | Rovná se (je stejné?) | `5 == 5` | `True` |
| `!=` | Nerovná se (je různé?) | `5 != 3` | `True` |
| `<` | Menší než | `3 < 5` | `True` |
| `>` | Větší než | `3 > 5` | `False` |
| `<=` | Menší nebo rovno | `5 <= 5` | `True` |
| `>=` | Větší nebo rovno | `3 >= 5` | `False` |

> ⚠️ **Pozor!** Nezaměňujte `=` (přiřazení) a `==` (porovnání)!
> - `x = 5` → *uloží* 5 do x
> - `x == 5` → *ptá se*, zda je x rovno 5

### Porovnání řetězců
Řetězce se porovnávají **lexikograficky** (podle abecedy, resp. podle ASCII/Unicode hodnot znaků):
- `"a" < "b"` → `True` (a je v abecedě před b)
- `"jedna" > "dva"` → `True` (protože `"j"` má vyšší ASCII hodnotu než `"d"`)
- Velká písmena mají **nižší** hodnotu než malá: `"A" < "a"` → `True`

### Operátor `!=` (nerovná se)
Opak operátoru `==` – vrací `True`, pokud se hodnoty **liší**:

```python
5 != 3       # True
"a" != "a"   # False
```

In [ ]:
# === Operátor == (rovná se?) ===
print("--- Rovnost (==) ---")
print("2 == 5:", 2 == 5)                 # False
print("5 == 5:", 5 == 5)                 # True
print("True == False:", True == False)    # False
print('"ahoj" == "ahoj":', "ahoj" == "ahoj")   # True
print('"jedna" == "dva":', "jedna" == "dva")     # False

# === Operátor != (nerovná se?) ===
print()
print("--- Nerovnost (!=) ---")
print("2 != 5:", 2 != 5)                 # True
print("5 != 5:", 5 != 5)                 # False

### Operátory `<`, `>`, `<=`, `>=`

In [ ]:
# === Porovnání čísel ===
print("--- Porovnání čísel ---")
print("2 < 5:", 2 < 5)       # True
print("5 > 5:", 5 > 5)       # False
print("2 >= 2:", 2 >= 2)     # True
print("3 <= 1:", 3 <= 1)     # False

# === Porovnání řetězců (lexikografické / podle ASCII) ===
print()
print("--- Porovnání řetězců ---")
print('"ahoj" < "ahoj":', "ahoj" < "ahoj")     # False (jsou stejné)
print('"abc" < "abd":', "abc" < "abd")           # True (c < d)
print('"jedna" > "dva":', "jedna" > "dva")       # True (j > d v ASCII)
print('"A" < "a":', "A" < "a")                   # True (velké písmeno < malé)

---

## ✍️ Zkrácený (složený) zápis operátorů

V programování velmi často potřebujeme **upravit hodnotu proměnné** na základě její aktuální hodnoty (např. přičíst, odečíst, …). Python pro to nabízí **zkrácený zápis**:

| Dlouhý zápis | Zkrácený zápis | Význam |
|---|---|---|
| `x = x + 5` | `x += 5` | Přičti 5 |
| `x = x - 3` | `x -= 3` | Odečti 3 |
| `x = x * 2` | `x *= 2` | Vynásob 2 |
| `x = x / 4` | `x /= 4` | Vyděl 4 |
| `x = x // 2` | `x //= 2` | Celočíselně vyděl 2 |
| `x = x % 3` | `x %= 3` | Zbytek po dělení 3 |
| `x = x ** 2` | `x **= 2` | Umocni na 2 |

### Příklad krok po kroku

In [ ]:
# Dlouhý zápis – funguje, ale je zbytečně ukecaný
var = 0
print("Počáteční hodnota:", var)
var = var + 2
print("Po var = var + 2:", var)     # 2

0
2


Zkrácený zápis – totéž, ale elegantněji:

In [ ]:
# Zkrácený zápis – totéž, ale kratší
print("Aktuální hodnota:", var)     # 2
var += 2    # přičte 2 → var = var + 2
print("Po var += 2:", var)          # 4

Funguje pro **všechny** matematické operátory:

In [ ]:
# Ukázka všech zkrácených zápisů
var = 10
print(f"Začínáme: var = {var}")

var += 5     # var = var + 5
print(f"var += 5  → {var}")     # 15

var -= 3     # var = var - 3
print(f"var -= 3  → {var}")     # 12

var *= 2     # var = var * 2
print(f"var *= 2  → {var}")     # 24

var //= 5    # var = var // 5
print(f"var //= 5 → {var}")    # 4

var **= 3    # var = var ** 3
print(f"var **= 3 → {var}")    # 64

var %= 10    # var = var % 10
print(f"var %= 10 → {var}")    # 4

2
-2
-4
-2.0
0.0


---

## 📐 Priorita (pořadí) operátorů

Stejně jako v matematice, i v Pythonu mají operátory **různou prioritu** – některé se vyhodnotí dříve než jiné.

### Pořadí od nejvyšší po nejnižší prioritu

| Priorita | Operátor | Popis |
|:---:|:---:|---|
| 1. (nejvyšší) | `()` | Závorky – vždy se vyhodnotí první |
| 2. | `**` | Mocnění |
| 3. | `+x`, `-x` | Unární plus/mínus (znaménko čísla) |
| 4. | `*`, `/`, `//`, `%` | Násobení, dělení, modulo |
| 5. | `+`, `-` | Sčítání, odčítání |
| 6. | `==`, `!=`, `<`, `>`, `<=`, `>=` | Porovnání |
| 7. (nejnižší) | `=`, `+=`, `-=`, … | Přiřazení |

### Příklad

```python
vysledek = 2 + 3 * 4      # = 2 + 12 = 14 (ne 20!)
vysledek = (2 + 3) * 4    # = 5 * 4 = 20 (závorky mění pořadí)
vysledek = 2 ** 3 ** 2     # = 2 ** 9 = 512 (mocnění se vyhodnocuje zprava!)
```

> 🎯 **Tip:** Pokud si nejste jisti prioritou, **používejte závorky**! Program bude čitelnější a vyhnete se chybám.

In [ ]:
# === Priorita operátorů ===
print("2 + 3 * 4   =", 2 + 3 * 4)       # 14 (násobení má přednost)
print("(2 + 3) * 4 =", (2 + 3) * 4)     # 20 (závorky mají přednost)

print()
print("2 ** 3 ** 2 =", 2 ** 3 ** 2)      # 512 (mocnění se vyhodnocuje zprava: 2^(3^2) = 2^9)
print("(2 ** 3) ** 2 =", (2 ** 3) ** 2)  # 64  ((2^3)^2 = 8^2)

print()
# Složitější výraz
print("10 - 2 * 3 + 4 / 2 =", 10 - 2 * 3 + 4 / 2)   # 10 - 6 + 2.0 = 6.0

---

# 📋 Shrnutí

| Téma | Co jsme se naučili |
|---|---|
| **Datové typy** | `int`, `float`, `str`, `bool` – Python typ určí automaticky |
| **`type()`** | Funkce pro zjištění datového typu |
| **Přetypování** | `int()`, `float()`, `str()`, `bool()` – převod mezi typy |
| **Aritmetické operátory** | `+`, `-`, `*`, `/`, `//`, `%`, `**` |
| **Porovnávací operátory** | `==`, `!=`, `<`, `>`, `<=`, `>=` → vrací `bool` |
| **Zkrácený zápis** | `+=`, `-=`, `*=`, `/=`, `//=`, `%=`, `**=` |
| **Priorita operátorů** | Závorky > mocnění > násobení/dělení > sčítání/odčítání > porovnání |
| **Float nepřesnost** | Počítače ukládají desetinná čísla s malou zaokrouhlovací chybou |

---

# ✏️ Cvičení

## Cvičení 1
Napište program, který **počítá se dvěma čísly** (uloženými v proměnných):
1. Sečte je
2. Součet umocní na třetí ($\text{součet}^3$)
3. Výsledek vydělí číslem $\pi$ (použijte `3.14`)
4. Na nový řádek vypíše: vstupní čísla, výsledek a datový typ výsledku

10 21 9487.579617834395 <class 'float'>


<details>
<summary>🔎 Výsledky pro kontrolu (klikni pro zobrazení)</summary>

Referenční řešení má **4 řádky kódu**.

| Vstup `a`, `b` | Očekávaný výstup |
|---|---|
| `4`, `3` | `4 3 109.23566878980891 <class 'float'>` |
| `1`, `1` | `1 1 2.5477707006369426 <class 'float'>` |
| `10`, `21` | `10 21 9487.579617834395 <class 'float'>` |

</details>

---

## Cvičení 2
Napište program, který:
1. Načte od uživatele **první celé číslo** a **vynásobí ho dvěma**
2. Načte **druhé celé číslo** a **celočíselně ho vydělí dvěma**
3. Vypíše obě zadaná čísla
4. Na další řádek vypíše oba výsledky výpočtů
5. Na třetí řádek vypíše, zda se tyto dva výsledky **rovnají** (`==`)

Zadejte první číslo:  


<details>
<summary>🔎 Výsledky pro kontrolu (klikni pro zobrazení)</summary>

Referenční řešení má **5 řádků kódu** (2× input, 3× print).

| Vstupy | Očekávaný výstup |
|---|---|
| `5`, `20` | `5 20` ↵ `10 10` ↵ `True` |
| `8`, `10` | `8 10` ↵ `16 5` ↵ `False` |
| `0`, `0` | `0 0` ↵ `0 0` ↵ `True` |

</details>

---

## Cvičení 3
Napište program, který:
1. Načte od uživatele **libovolné číslo** (i neceločíselné – tedy `float`)
2. Provede postupně tyto výpočty (každý krok používá výsledek předchozího):
   - Vypočítá **modulo třemi** (`% 3`)
   - Výsledek **umocní na druhou** (`** 2`)
   - Výsledek **vydělí jednou čtvrtinou** (`/ 0.25` neboli `/ (1/4)`)
   - Vypočítá **pátou mocninu** výsledku (`** 5`)
3. Vypíše **pod sebe všechny mezivýsledky**

zadejte prvníčíslo: 3.14
0.14000000000000012
0.019600000000000034
0.07840000000000014
2.9619676669542656e-06


<details>
<summary>🔎 Výsledky pro kontrolu (klikni pro zobrazení)</summary>

Referenční řešení má **6 řádků kódu** (1× input, 4× výpočet s print, nebo variace).

| Vstup | Mezivýsledky (4 řádky) |
|---|---|
| `1` | `1.0` ↵ `1.0` ↵ `4.0` ↵ `1024.0` |
| `3.14` | `0.14000000000000012` ↵ `0.019600000000000034` ↵ `0.07840000000000014` ↵ `2.9619676669542656e-06` |
| `0.5` | `0.5` ↵ `0.25` ↵ `1.0` ↵ `1.0` |

> 💡 Všimněte si nepřesnosti float u vstupu `3.14` – to je přesně ten problém, který jsme popisovali výše!

</details>

---

## 🌟 Bonusové cvičení 4
Napište program **kalkulačka**, který:
1. Načte od uživatele **dvě čísla** (mohou být desetinná)
2. Vypíše **všechny** výsledky aritmetických operací v přehledné formě:

```
=== KALKULAČKA ===
Číslo 1: 10
Číslo 2: 3
------------------
Součet:      13.0
Rozdíl:      7.0
Součin:      30.0
Podíl:       3.3333333333333335
Celočís. d.: 3.0
Zbytek:      1.0
Mocnina:     1000.0
------------------
```

> 💡 **Tip:** Použijte `float(input(...))` pro načtení čísel.

In [ ]:
# Bonusové cvičení 4: Váš kód sem 👇
